### ISSUE TITLE
Propose: from CLI we can support a profile concept, so user can switch to different chains apps easily. 

### ISSUE THREAD

**Issue by mave99a** [ArcBlock/forge-cli#11]
Propose: from CLI we can support a profile concept, so user can switch to different chains apps easily. 

e.g. we can use `forge start .`  it will look at a local config from the folder and put config, data all under that folder. 

This can be related to the proposal of `forge create-app` (#71) 

---
related to this:  https://github.com/ArcBlock/forge/issues/232

> In dev and test to develop a new dapps, user might need to run forge with different configuration back & forth.
> 
> Right now we put everything under ~/.forge-xxxx folder, which means, if I join a new chain, or start a new application, either the data will be messed up, or I have to clean up and resync every time.
> 
> So I propose to allow developer specify working folders easily, so they can easily switch back and forth.
> 
> Once this is supported, we can improve forge-cli to make the switch extremely easy and handy. It would be very helpful for developers.

Since this is already supported, I think we can do it from CLI side to make it easier.

**Comment by wangshijun** [comment]
@mave99a your proposal can be fulfilled with developer providing different forge config file path and forge release directory.

- forge config file path can be provided with `$FORGE_CONFIG`, or `--config-path`
- forge release dir can be provided with `$FORGE_RELEASE_DIR`, or `--release-dir`

And these 2 options can be combined to meet any developer needs to start forge with different config or release.

Here is my configuration to test latest forge master:

```
export FORGE_RELEASE_DIR=/Users/wangshijun/Develop/arcblock/forge/_build/staging/rel
```

When I test with latest forge release:

```
export FORGE_RELEASE_DIR=/Users/wangshijun/.forge_cli/release
```

When I need to start forge with simple custom config:


```
export FORGE_CONFIG=/Users/wangshijun/Develop/arcblock/forge-js/examples/simple/forge.toml
```

When I need to start forge with nodejs kvstore example config:

```
export FORGE_CONFIG=/Users/wangshijun/Develop/arcblock/forge-js/examples/kvstore/config/forge.toml
export FORGE_SDK_SETUP_SCRIPT=/Users/wangshijun/Develop/arcblock/forge-js/examples/kvstore/config/setup.js
```

**Comment by wangshijun** [comment]
However, your proposal seems provide better developer experience, on the data directory part. I will figure out how to acheive this.

**Comment by mave99a** [comment]
Yes, it will be easier for developer to switch between projects during the debug and development.  

This can be done together when you support `forge create-app`, we generate required config under that app directory. #35 

Imaging how GIT works: it simply look at .git directory, so it know where it is and which project you are working on. 

If we did support this mode, we should also warn user if the running instance isn't matching the current directory with configuration.

**Comment by mave99a** [comment]
I'd suggest to make our forge/examples  and final-chapter run in this way.  @wangshijun @tyrchen

**Comment by wangshijun** [comment]
@polunzh @mave99a @tyrchen we can extend `forge start` to support this feature:

We can get some inspiration from how node.js stores and resolves its dependencies in both global and local way. The approach is also similar to how git stores its data except that git has no global data.

The new `forge start` will store the following key info to a `profile`, which is actually a hidden folder named `.forge` located in some directory that can be resolved and accessed by forge-cli.

The key info included in `profile`:

- the forge version used to start the chain
- the config file of the chain
- the data folder and log folder of the chain

The lifecycle of a `profile` folder:

- Creating: `forge start .` or `forge start /path/to/folder`
- Reading: run `forge start` under a directory that contains `.forge`, simple validation should be done before use the `.forge` directory
- Destroying: user can delete `.forge` if it is not needed anymore 


When user run `forge start` under `~/Developer/arcblock/forge-demo-dapp/src`, forge-cli will behave differently based on the result of `.forge` folder resolution:

- `~/Developer/arcblock/forge-demo-app/.forge`: exist and valid, use this folder
- `~/Developer/arcblock/.forge`: exist and valid, use this folder
- both 1 and 2 exist, use 1
- neither 1 or 2 exist, fallback to `~/.forge_release`

To make the UX smooth with other forge subcommands, forge-cli should be smart enough to figure out a `profile context` before running any subcommand, then the `profile context` should be shared across all subcommands except the following:

```
forge install
forge download
forge wallet:create
forge help
forge protocol:*
forge tx:sign
forge tx:send
```

Besides, forge-cli should print out which profile is used if it's not the default one.

**Comment by wangshijun** [comment]
landed in forge-cli v0.36.0

https://github.com/ArcBlock/forge-tutorials/blob/master/src/forge-cli/index.md
